// @promisepending/logger.js/src/interfaces/ILoggerOption.js
"use strict";
/* eslint-disable no-unused-vars */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ELoggerLevelAlternateColors = exports.ELoggerLevelBaseColors = exports.ELoggerLevelNames = exports.ELoggerLevel = void 0;
var ELoggerLevel;
(function (ELoggerLevel) {
    ELoggerLevel[ELoggerLevel["INFO"] = 0] = "INFO";
    ELoggerLevel[ELoggerLevel["LOG"] = 0] = "LOG";
    ELoggerLevel[ELoggerLevel["WARN"] = 1] = "WARN";
    ELoggerLevel[ELoggerLevel["ALERT"] = 1] = "ALERT";
    ELoggerLevel[ELoggerLevel["ERROR"] = 2] = "ERROR";
    ELoggerLevel[ELoggerLevel["SEVERE"] = 2] = "SEVERE";
    ELoggerLevel[ELoggerLevel["FATAL"] = 3] = "FATAL";
    ELoggerLevel[ELoggerLevel["DEBUG"] = 4] = "DEBUG";
})(ELoggerLevel = exports.ELoggerLevel || (exports.ELoggerLevel = {}));
var ELoggerLevelNames;
(function (ELoggerLevelNames) {
    ELoggerLevelNames[ELoggerLevelNames["INFO"] = 0] = "INFO";
    ELoggerLevelNames[ELoggerLevelNames["WARN"] = 1] = "WARN";
    ELoggerLevelNames[ELoggerLevelNames["ERROR"] = 2] = "ERROR";
    ELoggerLevelNames[ELoggerLevelNames["FATAL"] = 3] = "FATAL";
    ELoggerLevelNames[ELoggerLevelNames["DEBUG"] = 4] = "DEBUG";
})(ELoggerLevelNames = exports.ELoggerLevelNames || (exports.ELoggerLevelNames = {}));
// color for text or colored background
var ELoggerLevelBaseColors;
(function (ELoggerLevelBaseColors) {
    ELoggerLevelBaseColors[ELoggerLevelBaseColors["#cc80ff"] = 0] = "#cc80ff";
    ELoggerLevelBaseColors[ELoggerLevelBaseColors["#ff8a1c"] = 1] = "#ff8a1c";
    ELoggerLevelBaseColors[ELoggerLevelBaseColors["#ff4a4a"] = 2] = "#ff4a4a";
    ELoggerLevelBaseColors[ELoggerLevelBaseColors["#ffffff"] = 3] = "#ffffff";
    ELoggerLevelBaseColors[ELoggerLevelBaseColors["#555555"] = 4] = "#555555";
})(ELoggerLevelBaseColors = exports.ELoggerLevelBaseColors || (exports.ELoggerLevelBaseColors = {}));
// color for text on colored background
exports.ELoggerLevelAlternateColors = [
    '#000000',
    '#000000',
    '#000000',
    '#ff0000',
    '#D4D4D4'
];
