// @twitchapis/twitch.js/src/client/connection/websocket/methods/001.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const structures_1 = require("../../../../structures");
class M001 {
    constructor(client) {
        this.client = client;
    }
    preLoad() {
        return Promise.resolve({ command: '001' });
    }
    execute(message) {
        return new Promise((resolve, reject) => {
            const command = message.command;
            if (!command || !command.channel)
                return reject('No parameters!');
            this.client.getLogger().debug('Logged in successfully as ' + command.channel);
            // Creates user collection
            this.client.user = new structures_1.BasicUserStructure(this.client, {
                username: command.channel,
                id: command.channel,
                self: true,
                userType: true,
                staff: false,
                admin: false,
                globalMod: false,
            });
            resolve();
        });
    }
}
exports.default = M001;
